File: ios/Sasquatsh/ViewModels/PlanningSessionViewModel.swift
import SwiftUI

@Observable
@MainActor
final class PlanningSessionViewModel {
    var session: PlanningSession?
    var isLoading = false
    var error: String?
    var actionMessage: String?

    private var services: ServiceContainer?

    func configure(services: ServiceContainer) {
        self.services = services
    }

    func loadSession(id: String) async {
        guard let services else { return }
        isLoading = true
        error = nil
        do {
            session = try await services.planning.getSession(id: id)
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    func submitResponse(cannotAttendAny: Bool, dateAvailability: [DateAvailabilityInput]) async {
        guard let services, let session else { return }
        error = nil
        do {
            let input = PlanningResponseInput(cannotAttendAny: cannotAttendAny, dateAvailability: dateAvailability)
            try await services.planning.submitResponse(sessionId: session.id, input: input)
            actionMessage = "Response submitted!"
            await loadSession(id: session.id)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func suggestGame(_ input: SuggestGameInput) async {
        guard let services, let session else { return }
        error = nil
        do {
            let suggestion = try await services.planning.suggestGame(sessionId: session.id, input: input)
            self.session?.gameSuggestions?.append(suggestion)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func voteForGame(suggestionId: String) async {
        guard let services, let session else { return }
        error = nil
        do {
            try await services.planning.voteForGame(sessionId: session.id, suggestionId: suggestionId)
            await loadSession(id: session.id)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func unvoteGame(suggestionId: String) async {
        guard let services, let session else { return }
        error = nil
        do {
            try await services.planning.unvoteGame(sessionId: session.id, suggestionId: suggestionId)
            await loadSession(id: session.id)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func removeSuggestion(suggestionId: String) async {
        guard let services, let session else { return }
        error = nil
        do {
            try await services.planning.removeSuggestion(sessionId: session.id, suggestionId: suggestionId)
            self.session?.gameSuggestions?.removeAll { $0.id == suggestionId }
        } catch {
            self.error = error.localizedDescription
        }
    }

    func finalize(selectedDateId: String?, selectedGameId: String?) async -> String? {
        guard let services, let session else { return nil }
        error = nil
        do {
            let result = try await services.planning.finalizeSession(sessionId: session.id, selectedDateId: selectedDateId, selectedGameId: selectedGameId)
            actionMessage = result.message
            return result.eventId
        } catch {
            self.error = error.localizedDescription
            return nil
        }
    }

    func addItem(name: String, category: ItemCategory, quantity: Int?) async {
        guard let services, let session else { return }
        error = nil
        do {
            let input = AddPlanningItemInput(itemName: name, itemCategory: category, quantityNeeded: quantity)
            let item = try await services.planning.addItem(sessionId: session.id, input: input)
            if self.session?.items == nil {
                self.session?.items = []
            }
            self.session?.items?.append(item)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func claimItem(itemId: String) async {
        guard let services, let session else { return }
        error = nil
        do {
            try await services.planning.claimItem(sessionId: session.id, itemId: itemId)
            await loadSession(id: session.id)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func unclaimItem(itemId: String) async {
        guard let services, let session else { return }
        error = nil
        do {
            try await services.planning.unclaimItem(sessionId: session.id, itemId: itemId)
            await loadSession(id: session.id)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func removeItem(itemId: String) async {
        guard let services, let session else { return }
        error = nil
        do {
            try await services.planning.removeItem(sessionId: session.id, itemId: itemId)
            self.session?.items?.removeAll { $0.id == itemId }
        } catch {
            self.error = error.localizedDescription
        }
    }

    func updateSettings(tableCount: Int?) async {
        guard let services, let session else { return }
        error = nil
        do {
            try await services.planning.updateSettings(sessionId: session.id, tableCount: tableCount)
            self.session?.tableCount = tableCount
            actionMessage = tableCount != nil && tableCount! >= 2
                ? "Multi-table enabled with \(tableCount!) tables"
                : "Multi-table disabled"
        } catch {
            self.error = error.localizedDescription
        }
    }

    func cancel() async {
        guard let services, let session else { return }
        error = nil
        do {
            try await services.planning.cancelSession(sessionId: session.id)
            actionMessage = "Session cancelled"
            await loadSession(id: session.id)
        } catch {
            self.error = error.localizedDescription
        }
    }
}

File: android/app/src/main/java/com/sasquatsh/app/viewmodels/PlanningSessionViewModel.kt
package com.sasquatsh.app.viewmodels

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sasquatsh.app.models.AddPlanningItemInput
import com.sasquatsh.app.models.DateAvailabilityInput
import com.sasquatsh.app.models.ItemCategory
import com.sasquatsh.app.models.PlanningResponseInput
import com.sasquatsh.app.models.PlanningSession
import com.sasquatsh.app.models.ScheduleEntry
import com.sasquatsh.app.models.SuggestGameInput
import com.sasquatsh.app.services.PlanningService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class PlanningSessionUiState(
    val session: PlanningSession? = null,
    val isLoading: Boolean = false,
    val error: String? = null,
    val actionMessage: String? = null
)

@HiltViewModel
class PlanningSessionViewModel @Inject constructor(
    private val planningService: PlanningService,
    savedStateHandle: SavedStateHandle
) : ViewModel() {

    private val _uiState = MutableStateFlow(PlanningSessionUiState())
    val uiState: StateFlow<PlanningSessionUiState> = _uiState.asStateFlow()

    private val sessionId: String? = savedStateHandle["sessionId"]

    init {
        sessionId?.let { loadSession(it) }
    }

    fun loadSession(id: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            try {
                val session = planningService.getSession(id)
                _uiState.update { it.copy(session = session, isLoading = false) }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage, isLoading = false) }
            }
        }
    }

    fun submitResponse(cannotAttendAny: Boolean, dateAvailability: List<DateAvailabilityInput>) {
        val session = _uiState.value.session ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                val input = PlanningResponseInput(
                    cannotAttendAny = cannotAttendAny,
                    dateAvailability = dateAvailability
                )
                planningService.submitResponse(session.id, input)
                _uiState.update { it.copy(actionMessage = "Response submitted!") }
                loadSession(session.id)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun suggestGame(input: SuggestGameInput) {
        val session = _uiState.value.session ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                val suggestion = planningService.suggestGame(session.id, input)
                _uiState.update { state ->
                    val currentSession = state.session ?: return@update state
                    val suggestions = (currentSession.gameSuggestions ?: emptyList()) + suggestion
                    state.copy(session = currentSession.copy(gameSuggestions = suggestions))
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun voteForGame(suggestionId: String) {
        val session = _uiState.value.session ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                planningService.voteForGame(session.id, suggestionId)
                loadSession(session.id)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun unvoteGame(suggestionId: String) {
        val session = _uiState.value.session ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                planningService.unvoteGame(session.id, suggestionId)
                loadSession(session.id)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun removeSuggestion(suggestionId: String) {
        val session = _uiState.value.session ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                planningService.removeSuggestion(session.id, suggestionId)
                _uiState.update { state ->
                    val currentSession = state.session ?: return@update state
                    val suggestions = currentSession.gameSuggestions?.filter { it.id != suggestionId }
                    state.copy(session = currentSession.copy(gameSuggestions = suggestions))
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun finalize(selectedDateId: String?, selectedGameId: String?, onEventCreated: (String) -> Unit) {
        val session = _uiState.value.session ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                val result = planningService.finalizeSession(session.id, selectedDateId, selectedGameId)
                _uiState.update { it.copy(actionMessage = result.message) }
                onEventCreated(result.eventId)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun addItem(name: String, category: ItemCategory, quantity: Int?) {
        val session = _uiState.value.session ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                val input = AddPlanningItemInput(
                    itemName = name,
                    itemCategory = category,
                    quantityNeeded = quantity
                )
                val item = planningService.addItem(session.id, input)
                _uiState.update { state ->
                    val currentSession = state.session ?: return@update state
                    val items = (currentSession.items ?: emptyList()) + item
                    state.copy(session = currentSession.copy(items = items))
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun claimItem(itemId: String) {
        val session = _uiState.value.session ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                planningService.claimItem(session.id, itemId)
                loadSession(session.id)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun unclaimItem(itemId: String) {
        val session = _uiState.value.session ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                planningService.unclaimItem(session.id, itemId)
                loadSession(session.id)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun removeItem(itemId: String) {
        val session = _uiState.value.session ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                planningService.removeItem(session.id, itemId)
                _uiState.update { state ->
                    val currentSession = state.session ?: return@update state
                    val items = currentSession.items?.filter { it.id != itemId }
                    state.copy(session = currentSession.copy(items = items))
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun scheduleSessions(assignments: Map<Int, String>) {
        val session = _uiState.value.session ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                val schedule = assignments.map { (tableNum, gameId) ->
                    ScheduleEntry(
                        suggestionId = gameId,
                        tableNumber = tableNum,
                        slotIndex = 0
                    )
                }
                planningService.scheduleSessions(session.id, schedule)
                loadSession(session.id)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun updateSettings(tableCount: Int?) {
        val session = _uiState.value.session ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                planningService.updateSettings(session.id, tableCount)
                _uiState.update { state ->
                    val currentSession = state.session ?: return@update state
                    val message = if (tableCount != null && tableCount >= 2) {
                        "Multi-table enabled with $tableCount tables"
                    } else {
                        "Multi-table disabled"
                    }
                    state.copy(
                        session = currentSession.copy(tableCount = tableCount),
                        actionMessage = message
                    )
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun cancel() {
        val session = _uiState.value.session ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                planningService.cancelSession(session.id)
                _uiState.update { it.copy(actionMessage = "Session cancelled") }
                loadSession(session.id)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun clearActionMessage() {
        _uiState.update { it.copy(actionMessage = null) }
    }
}
